### PR TITLE
Ignore panic functions in backtraces. Print inlined functions on Windows

### PR DESCRIPTION
**Pull request by Zoxc**
For the following program,
```rust
fn main() {
    panic!("hello world");
}
```
we now get this stack trace:
```
thread 'main' panicked at 'hello world', test.rs:2:4
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: test::main
             at .\test.rs:2
```
instead of this one:
```
thread 'main' panicked at 'hello world', test.rs:2:4
stack backtrace:
   0: std::sys_common::backtrace::_print
             at C:\projects\rust\src\libstd\sys_common\backtrace.rs:92
   1: std::panicking::default_hook::{{closure}}
             at C:\projects\rust\src\libstd\panicking.rs:380
   2: std::panicking::default_hook
             at C:\projects\rust\src\libstd\panicking.rs:397
   3: std::panicking::rust_panic_with_hook
             at C:\projects\rust\src\libstd\panicking.rs:577
   4: std::panicking::begin_panic<str*>
             at C:\projects\rust\src\libstd\panicking.rs:538
   5: test::main
             at .\test.rs:2
   6: panic_unwind::__rust_maybe_catch_panic
             at C:\projects\rust\src\libpanic_unwind\lib.rs:99
   7: std::rt::lang_start
             at C:\projects\rust\src\libstd\rt.rs:52
   8: __scrt_common_main_seh
             at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:283
   9: BaseThreadInitThunk
```

This introduces a `str` macro fragment used to redirect `panic!` called with string literals to a non-generic function. This redirection is also an optimization (and matches what libcore does).

### CODE REVIEW

**Comment by kennytm** [comment]
cc #40264 (@Yamakaky)

cc #35625 (`:literal` macro specifier)

**Comment by Yamakaky** [comment]
SGTM!

File: src/libsyntax/ext/tt/macro_parser.rs
Comment (petrochenkov) [inline src/libsyntax/ext/tt/macro_parser.rs]: All the previous matchers went through RFC process.
Well, at least I'm pretty sure this addition should not be hidden in a PR named "Ignore panic functions in backtraces".

**Comment by Zoxc** [comment]
I removed the `str` macro fragment specifier from this PR.

This does not seem to work on OS X, probably because the backtraces are incorrect there.

**Comment by Zoxc** [comment]
I changed how the implemented worked. Now panic functions passes a pointer to themselves into the panic handler. Backtraces then ignore functions in the backtrace before that pointer. Any functions after `mark_start` (calls main and thread functions) are also emitted from backtraces.

File: src/libtest/lib.rs
Comment (alexcrichton) [inline src/libtest/lib.rs]: This presumably no longer works now that the libstd implementation was updated?

Comment (Zoxc) [reply src/libtest/lib.rs]: Yeah this is now broken. We could work around it by exporting `mark_start` as an unstable item.

File: src/libstd/sys_common/backtrace.rs
Comment (Zoxc) [reply src/libstd/sys_common/backtrace.rs]: It does not. It just prevents inlining, not the removal of stack frames.

File: src/libstd/panicking.rs
Comment (Zoxc) [reply src/libstd/panicking.rs]: Probably just a leftover from the earlier version.

File: src/libstd/sys_common/backtrace.rs
Comment (Zoxc) [reply src/libstd/sys_common/backtrace.rs]: I think so. An earlier version had an array of functions, which needed casts.

File: src/libcore/panicking.rs
Comment (Zoxc) [reply src/libcore/panicking.rs]: I think passing locals by reference also prevents tail call optimization. I'm not really sure that this or the `asm!` trick works for function which never return. Currently these do not get tail call optimization applied, but I don't see a reason for that. Perhaps LLVM skips that optimization on cold functions?

File: src/libstd/sys_common/backtrace.rs
Comment (kennytm) [reply src/libstd/sys_common/backtrace.rs]: Instead of `asm!`, does `std::sync::atomic::compiler_fence()` work?

Comment (Zoxc) [reply src/libstd/sys_common/backtrace.rs]: It shouldn't.

**Comment by Zoxc** [comment]
Now this also prints inlined functions in backtraces on Windows.

@alexcrichton I'd like this to get merged in time for the next beta so I can get more complete backtraces in rustc.

cc @retep998

**Review by bjorn3 — CHANGES_REQUESTED** [review]

**Comment by kennytm** [comment]
Ping @alexcrichton (also @retep998) is there any unresolved questions for this PR?

**Comment by kennytm** [comment]
Triage ping @alexcrichton! There are some responses from @Zoxc on https://github.com/rust-lang/rust/pull/45637#pullrequestreview-73517560 and https://github.com/rust-lang/rust/pull/45637#issuecomment-344670053

File: src/libcore/panicking.rs
Comment (alexcrichton) [reply src/libcore/panicking.rs]: I'm a little wary to have this and rely on this, are we sure that this doesn't get optimized away even with LTO? If we leave this in can it at least be documented in the source as to why it's a reference?

Comment (alexcrichton) [reply src/libcore/panicking.rs]: My guess for why this isn't optimized is because `fmt::Arguments` is a pretty big struct that's copied around the stack, and LLVM is probably conservative with the tail-call for now. Despite that though this seems like something that's pretty brittle to rely on?

File: src/libtest/lib.rs
Comment (alexcrichton) [inline src/libtest/lib.rs]: Does this need to be either `#[inline]` or `#[inline(always)]`? If so, can it be documented as to why?

**Comment by alexcrichton** [comment]
Nice on the windows bits! Could those perhaps be moved to a separate PR? They seem unambiguously good to me while I'm still pretty uneasy on the other parts.

It seems like relying on the exact addresses of functions is quite brittle? I'm worried that this will cause regressions over time as we continue to upgrade LLVM. Is there perhaps something we can do to make this more bullet-proof rather than "insert something and pray llvm doesn't optimize it"?

File: src/libtest/lib.rs
Comment (Zoxc) [reply src/libtest/lib.rs]: I don't think the attribute is needed here.

**Comment by Zoxc** [comment]
I don't think there's anything we can do to make it bulletproof without modifying LLVM. If it breaks, it will just return to the status quo, so I don't think that's a good reason to avoid landing this. Backtraces are unfortunately best effort due to other reasons anyway.

**Comment by alexcrichton** [comment]
I'm personally sort of wary for that though in that this if we go down this road it's a commitment for us to make that we will receive, review, and maintain patches to fix this functionality as it regresses over time. In that sense I'm not comfortable myself r+'ing this but would want a larger sign-off from perhaps @rust-lang/libs on this functionality.

**Comment by Zoxc** [comment]
Actually LLVM supports marking function calls as `notail`, so if we want to make this reliable we could add an attribute to mark function calls as `notail`.

**Comment by alexcrichton** [comment]
Oh nice! Could the `notail` attribute be leveraged in that case? That seems like the most robust solution available to us, right?

**Comment by alexcrichton** [comment]
This was discussed at the libs triage meeting today and the conclusion was that we'd be comfortable landing this with the `notail` attribute added here, @Zoxc would you be up for doing that?

**Comment by Zoxc** [comment]
It sounds like that is going to require an RFC. I don't really want to write one.

**Comment by sfackler** [comment]
I don't think it requires an RFC to use a new LLVM feature in rust-lang/rust internals.

**Comment by Zoxc** [comment]
It would be a new language feature though.

**Comment by sfackler** [comment]
It would require an RFC to *stabilize* but not to use as an internal implementation detail.

**Comment by alexcrichton** [comment]
Oh yeah sorry to be clear things like that tend to not need RFCs, for example lots of the allocator fiddly bits weren't ever RFC'd or specified, they're just various attributes the compiler reads.

**Comment by kennytm** [comment]
Hi @Zoxc just checking if you  are still working on this PR 🙂.

Creating an internal attribute doesn't require an RFC or FCP (there are tons of such attributes like `#[rustc_if_this_changed]` or `#[sanitizer_runtime]`, check `src/libsyntax/feature_gate.rs`), so you are free to go ahead with `#[llvm_notail]` or something like this.

**Comment by Zoxc** [comment]
We don't have attributes on function calls in MIR. I guess we could use a `#[notail_if_called]` attribute instead.

**Comment by Zoxc** [comment]
@alexcrichton I added the attribute.

@eddyb Would you look at the trans changed in the last commit?

File: src/rustllvm/RustWrapper.cpp
Comment (eddyb) [inline src/rustllvm/RustWrapper.cpp]: Can you make a separate function that calls `setTailCallKind`?

**Comment by eddyb** [comment]
@Zoxc I think the name of the attribute should be `notail_when_called` or `never_tail_call` etc.
The "if" in `notail_if_called` seems out of place, at least to me.

File: src/rustllvm/RustWrapper.cpp
Comment (Zoxc) [reply src/rustllvm/RustWrapper.cpp]: What is the benefit of that?

Comment (eddyb) [reply src/rustllvm/RustWrapper.cpp]: Not having random boolean arguments in the API, both in C++ and Rust.

Comment (Zoxc) [reply src/rustllvm/RustWrapper.cpp]: I added a separate function to mark CallInsts as notail.

File: src/libcore/panicking.rs
Comment (alexcrichton) [inline src/libcore/panicking.rs]: Could you add a comment for what this attribute is doing?

File: src/libstd/panicking.rs
Comment (alexcrichton) [inline src/libstd/panicking.rs]: Could you add comment as well to functions in this crate for why this attribute is applied?

**Comment by alexcrichton** [comment]
Looks good to me, thanks! A few comments and a fixup to some indentation and otherwise looks ready to go.

File: src/librustc_trans/mir/block.rs
Comment (eddyb) [inline src/librustc_trans/mir/block.rs]: This *should probably* be `invokeret`.

File: src/test/ui/feature-gate-notail_when_called.rs
Comment (eddyb) [inline src/test/ui/feature-gate-notail_when_called.rs]: Missing newline at the end of file.

File: src/librustc_trans/mir/block.rs
Comment (Zoxc) [reply src/librustc_trans/mir/block.rs]: That would be bad, since `InvokeInst` is unrelated to `CallInst` and does not have a `notail` attribute.

Comment (eddyb) [reply src/librustc_trans/mir/block.rs]: They're both `CallSite` though?

Comment (Zoxc) [reply src/librustc_trans/mir/block.rs]: I don't think that has a `setTailCallKind` method.

File: src/rustllvm/RustWrapper.cpp
Comment (eddyb) [inline src/rustllvm/RustWrapper.cpp]: Okay, then I'd make this `if (auto CI = dyn_cast<CallInst>(unwrap(Call))) CI->...` just so you can pass an invoke instruction.

File: src/librustc_trans/mir/block.rs
Comment (eddyb) [reply src/librustc_trans/mir/block.rs]: Oh, my bad, I just checked again and that is indeed that case. That's a big troubling, because an `invoke` can become a `call` through optimizations :/.

Comment (Zoxc) [reply src/librustc_trans/mir/block.rs]: We should probably panic if we need to add notail on invokes then. I'll do that to check that we don't emit any invokes here.

Comment (Zoxc) [reply src/librustc_trans/mir/block.rs]: It seems like we do emit invokes here.

Comment (eddyb) [reply src/librustc_trans/mir/block.rs]: I guess the attribute should require non-Rust ABI which implies `nounwind`?

Comment (Zoxc) [reply src/librustc_trans/mir/block.rs]: Isn't `invoke` only used when you need cleanup? Normal `call` instructions can also unwind and should be used here

Comment (eddyb) [reply src/librustc_trans/mir/block.rs]: Right, unwinding + cleanup.

**Comment by Zoxc** [comment]
@alexcrichton It turns out `notail` cannot be applied to `invoke` instructions in LLVM, and LLVM can convert `invoke` instructions into `call` instructions (without `notail`).

**Comment by alexcrichton** [comment]
@Zoxc that's unfortunate :(

Would it be possible to land the fixes for inlined functions on Windows and perhaps follow-up with an upstream bug with LLVM to add this feature?

**Comment by kennytm** [comment]
Hi @Zoxc, is this PR currently blocked by #47252, and is there a workaround for `notail invoke` or a LLVM bug report we can follow as mentioned in https://github.com/rust-lang/rust/pull/45637#issuecomment-355046950? Thanks!

**Comment by Zoxc** [comment]
There isn't any workaround nor have I filed a LLVM issue.

I should still split out the fixes to mark the start of backtraces though, which addresses https://github.com/rust-lang/rust/issues/47429.

**Comment by shepmaster** [comment]
Ping from triage, @Zoxc ! Have you "split out the fixes to mark the start of backtraces"? If so, should this PR be closed?

**Pull request closed without merging**
